numpyro/hmc_util.py

```diff
@@ -553,13 +554,27 @@ def logp(d, val):
     return log_joint, model_trace
 
 
-def potential_energy(model, model_args, model_kwargs):
-    return lambda params: - jax.partial(log_density, model, model_args, model_kwargs)(params)[0]
+def potential_energy(model, model_args, model_kwargs, transforms):
+    def _potential_energy(params):
+        params_constrained = {k: transforms[k](v) for k, v in params.items()}
+        log_joint = jax.partial(log_density, model, model_args, model_kwargs)(params_constrained)[0]
+        for name, t in transforms.items():
+            log_joint = log_joint + np.sum(t.log_abs_det_jacobian(params[name], params_constrained[name]))
+        return - log_joint
+
+    return _potential_energy
+
+
+def transform_fn(transforms, params, invert=False):
+    return {k: transforms[k](v) if not invert else transforms[k].inv(v)
+            for k, v in params.items()}
```

Comment:
The logic of transforms and log_ab_det_jac looks correct to me! It might be worth to add a comment (to avoid the confusion with transforms belong to in Transform class) about the convention that: transform is from constrained to unconstrained, and inv transform is from unconstrained to constrained. But this can be added to docs later instead.

Reply:
Good point. Will remember to add this to the docs!